### PR TITLE
Remove extra menu separators

### PR DESCRIPTION
**Pull request by codebytere**
Removes leading and trailing menu separators and resolves https://github.com/electron/electron/issues/5869. 

Tested manually on MacOS, Windows, and Linux

To-Do:
- [x] Add specs

### CODE REVIEW

File: lib/browser/api/menu.js [hunk 1]
```diff
@@ -262,6 +264,18 @@ function indexToInsertByPosition (items, position) {
   return (query in queries) ? queries[query](idx) : idx
 }
 
+function removeExtraSeparators (template) {
+  let visiblePrev
+  const visibleItems = template.filter(el => el.visible !== false)
+
+  return visibleItems.filter((el, idx, array) => {
+    const meetsDeleteConditions = !visiblePrev || idx === array.length - 1 || array[idx + 1].type === 'separator'
+    const toDelete = el.type === 'separator' && meetsDeleteConditions
```
Comment (felixrieseberg) [inline lib/browser/api/menu.js]: This is _a little_ hard to scan. I think that a `reduce` would work wonders here.

Comment (ckerr) [reply lib/browser/api/menu.js]: I agree with felix this is a little hard to scan. I think the reasons are:
* `meetsDeleteConditions` and `toDelete` are confusingly named, e.g. how can you meet delete conditions if type isn't `separator`?
* `visibleItems` requires a little thought before you can tell how the loop will behave in edge cases

I'm not thinking functionally enough to see how a `reduce` would work cleanly here and would love enlightenment on this, no snark intended :smile: 

After poking around with a few approaches, the clearest I came up with was (untested):

```js
function removeExtraSeparators(items) {
  // remove invisible items 
  items = items.filter(e => e.visible !== false)
  // fold adjacent separators together
  items = items.filter((e, idx, arr) => e.type !== 'separator' || idx==0 || arr[idx-1].type!=='separator')
  // remove edge separators
  return items.filter((e, idx, arr) => e.type !== 'separator' || (idx!=0 && idx!=arr.length-1))
}
```

Marking as Request Changes for cleanup to this section, but the above code is just a suggestion, not a requirement

File: spec/api-menu-spec.js [hunk 2]
```diff
@@ -118,14 +121,36 @@ describe('Menu module', () => {
...
+            label: 'c'
+          }, {
+            type: 'separator'
+          }
+        ])
+
+        assert.equal(menu.items.length, 3)
+        assert.equal(menu.items[0].label, 'a')
+        assert.equal(menu.items[1].label, 'b')
+        assert.equal(menu.items[2].label, 'c')
```
Comment (ckerr) [inline spec/api-menu-spec.js]: Another good edge case to test for would be [ sep, sep, a, b, c, sep, sep ] -> [ a, b, c ]
